misk-metrics/src/test/kotlin/misk/metrics/v3/PeakGaugeTest.kt

```diff
@@ -0,0 +1,101 @@
+package misk.metrics.v3
+
+import io.micrometer.core.instrument.Clock
+import io.micrometer.core.instrument.Tag
+import io.micrometer.prometheus.PrometheusConfig
+import io.micrometer.prometheus.PrometheusMeterRegistry
+import io.prometheus.client.CollectorRegistry
+import misk.metrics.v2.PeakGauge
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+internal class PeakGaugeTest {
+
+  /**
+   * Verifies that a peak gauge records the correct maximum value and resets to the initial value
+   * after [PeakGauge.getAndClear]
+   */
+  @Test fun `records maximum and resets on getAndClear`() {
+    val gauge = Metrics.Companion.PeakGauge()
+
+    // 0 is our implicit initial value
+    assertThat(gauge.getAndClear()).isEqualTo(0.0)
+
+    // We can read a value after setting it
+    gauge.record(37.0)
+    assertThat(gauge.getAndClear()).isEqualTo(37.0)
+
+    // Clearing the value should reset it to the initial value
+    assertThat(gauge.getAndClear()).isEqualTo(0.0)
+
+    // A read after multiple sets returns the peak value
+    gauge.record(10.0)
+    gauge.record(5.0)
+    gauge.record(15.0)
+    gauge.record(20.0)
+    gauge.record(1.0)
+
+    assertThat(gauge.getAndClear()).isEqualTo(20.0)
+  }
+
+  /**
+   * Verifies that [PeakGauge.collect] returns correct information and from and resets each
+   * [PeakGauge].
+   */
+  @Test fun `collection returns correct information and resets each child`() {
+    val registry = CollectorRegistry(true)
+    val metrics = Metrics(PrometheusMeterRegistry(PrometheusConfig.DEFAULT, registry, Clock.SYSTEM))
+
+    val gauge1 = Metrics.Companion.PeakGauge()
+    metrics.gauge(
+      "some_name", gauge1::getAndClear, "some help text", listOf(
```

Comment:
This seems a little unwieldy. Should there be a `metrics.v3.Metrics#peakGauge(name: String, description: String, tags: List<Tag>)` method?

That would help migrate from v2's peakGauge as well.

Reply:
Unfortunately I don't think so (but happy for solutions). Micrometer gauges expect to have an object they can observe, and making many copies of the same thing is likely to result in errors.

PeakGauge is used only in one or two places in cash, and it's overall necessity should be lower with micrometer and some adjustments to how we do metrics (which should come with dumping bucketing if I understood the concerns correctly)

Reply:
What about something like this?

https://github.com/cashapp/misk/compare/lovelace/MiskMetrics/MetricsV3...keefer.20231023.keefer.metrics-v3-peak-gauge-suggestion?expand=1